working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+  - does not impose any restrictions on objects, you could associate a
+    finalization action with;
+  - does not impose any restrictions on finalization actions, but invokes them
+    asynchronously;
+  - is implementable on the Web.
+- `NativeFinalizationRegistry` which:
+  - provides stronger guarantees around promptness of finalization;
+  - guarantees that finalization actions are invoked when isolate is shutting
+  down;
+  - is limited to objects which implement `Finalizable` interface;
+  - imposes restrictions on finalization actions to allow calling these actions
+  outside of Dart universe.
+
+The following classes are added to `dart:core` library
+
+```dart
+/// A register of objects which may invoke a callback when those objects
+/// become inaccessible.
+///
+/// The register allows objects to be registered,
+/// and when those objects become inaccessible to the program.
+/// the callback passed to the register's constructor *may* be called
+/// with the registration token associated with the object.
+///
+/// No promises are made that the callback will ever be called,
+/// only that *if* it is called with a finalization token as argument,
+/// at least one object registered in the registry with that finalization token
+/// is no longer accessible to the program.
+///
+/// If the same object is registered in multiple finalization registries,
+/// or registered multiple times in a single registry,
+/// and the object becomes inaccessible to the program,
+/// then any number of those registrations may trigger their associated
+/// callback. It will not necessarily be all or none of them.
+///
+/// Finalization callbacks will happen as *events*, not during execution of
+/// other code and not as a microtask, but as high-level events similar to
+/// timer events.
+abstract class FinalizationRegistry<F extends Object, FT> {
+  /// Creates a finalization registry with the given finalization callback.
+  external factory FinalizationRegistry(
+    void Function(FT finalizationToken) callback);
+
+  /// Registers [value] for a finalization callback.
+  ///
+  /// When [value] is no longer accessible to the program,
+  /// the registry *may* call its callback function with [finalizationToken]
+  /// as argument.
+  ///
+  /// The [value] and [unregisterToken] arguments do not count towards those
+  /// objects being accessible to the program. Both must be non-[num],
+  /// non-[String], non-[bool], non-[Pointer], non-[Struct] and
+  /// non-[Null] values.
+  ///
+  /// Multiple objects may be registered with the same finalization token,
+  /// and the same object may be registered multiple times with different,
+  /// or the same, finalization token.
+  ///
+  /// The callback may be called at most once per registration, and not
+  /// for registrations which have been unregistered since they were registered.
+  void register(F value, FT finalizationToken, {Object? unregisterToken});
```

Comment:
nit: change
`{Object? unregisterToken}`
to
`{Object unregisterToken}` or `{Object unregisterToken = noUnregisterToken}`
as undefined is not a valid unregisterToken, there is no real point in allowing null unregisterTokens and Dart2JS may unpredictably use `null` or `undefined` for `null`. I believe `null` is a valid unregister token in JS but `undefined` is not.

Reply:
It should probably be `{Object? unregisterToken}` with `null` meaning not having any unregister token for that registration (it's optional, meaning it's possible to not have one).
The `unregister` should take `Object`.